platform/admin/core/admin-core-appservice/src/main/java/org/codice/ddf/admin/application/service/migratable/ProfileMigratable.java

```diff
@@ -0,0 +1,280 @@
+import com.google.common.collect.ImmutableMap;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import org.apache.commons.io.IOUtils;
+import org.apache.karaf.features.FeaturesService;
+import org.boon.json.JsonFactory;
+import org.boon.json.ObjectMapper;
+import org.codice.ddf.admin.application.service.ApplicationService;
+import org.codice.ddf.migration.ExportMigrationContext;
+import org.codice.ddf.migration.ImportMigrationContext;
+import org.codice.ddf.migration.Migratable;
+import org.codice.ddf.migration.MigrationException;
+import org.codice.ddf.migration.MigrationReport;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Class used to migrate the state of applications, features, and bundles. */
+public class ProfileMigratable implements Migratable {
+
+  private static final Charset DEFAULT_CHARSET = StandardCharsets.UTF_8;
+
+  public static final String PROFILE_ID = "id";
+
+  public static final String PROFILE_NAME = "name";
+
+  public static final String PROFILE_VERSION = "version";
+
+  public static final String PROFILE_DESCRIPTION = "description";
+
+  public static final String PROFILE_REPOSITORY = "repository";
+
+  public static final String PROFILE_URI = "uri";
+
+  public static final String PROFILE_LOCATION = "location";
+
+  public static final String PROFILE_STATE = "state";
+
+  public static final String PROFILE_START = "start";
+
+  public static final String PROFILE_REGION = "region";
+
+  public static final String PROFILE_APPS = "apps";
+
+  public static final String PROFILE_FEATURES = "features";
+
+  public static final String PROFILE_BUNDLES = "bundles";
+
+  /**
+   * Holds a retry count.
+   *
+   * <p>Because it was seen with Karaf that sometimes installing or starting a feature or a bundle
+   * might have side effects and install or start others which were not supposed to in the original
+   * system. As such, we are going to attempt multiple times for each operations such that we can
+   * account for side effects of other operations. In addition, because we are unable to determine a
+   * proper order to process all features or bundles, it is possible that we end up attempting to
+   * install a feature for which the dependencies have not yet been installed. We are therefore
+   * forced to attempt multiple times hoping that dependencies are properly being handled by the
+   * previous pass. Failures will be reported only on the last attempt. Retries will automatically
+   * stop when all apps, features, and apps are determined to be as they were at export time.
+   */
+  @SuppressWarnings("PMD.DefaultPackage" /* designed to be used by TaskList within this package */)
+  static final int RETRY_COUNT = 5;
```

Comment:
❓ Looks like there's no easy way to pass this down to the `TaskList` constructor instead of making it package-private, but having `TaskList` depend on this class seems odd. Should we have two separate constants instead or put the constant in the `TaskList` class?

Reply:
- Definitely not a fan of having two constants to mean the same thing ... harder to maintain
- Moving it to the TaskList creates the same problem as ProfileMigratable would now depend on TaskList when it does not

The other possibility is to simply move it to a different class just to hold a constant which I think becomes an overkill.

I chose this class as it is the center point of this project so it seemed a good place to put a common constant for now.